app/src/org/commcare/tts/TextToSpeechConverter.kt

```diff
@@ -0,0 +1,191 @@
+     * Good to call this from onDestroy().
+     */
+    fun shutDown() {
+        mTextToSpeech?.let {
+            it.shutdown()
+        }
+        mInitialized = false
+    }
+
+    private fun speakInternal(tts: TextToSpeech, text: String, queueMode: Int = TextToSpeech.QUEUE_FLUSH) {
+        val utteranceId = System.currentTimeMillis().toString()
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
+            tts.speak(text, queueMode, null, utteranceId)
+        } else {
+            val params = HashMap<String, String>()
+            params[TextToSpeech.Engine.KEY_PARAM_UTTERANCE_ID] = utteranceId
+            tts.speak(text, queueMode, params)
+        }
+    }
+
+    private fun isTextLong(text: String): Boolean {
+        // TTS can only speak 4000 characters at max at a time.
+        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2) {
+            text.length > TextToSpeech.getMaxSpeechInputLength()
+        } else {
+            text.length > MAX_TEXT_LENGTH
+        }
+    }
+
+    /**
+     * Initializes the Text-To-Speech engine
+     */
+    private fun initialize(context: Context, text: String) {
+        mTextToSpeech = TextToSpeech(context, TextToSpeech.OnInitListener { status ->
+            if (status == TextToSpeech.SUCCESS) {
+                // set language and speak
+                mTextToSpeech?.let { tts ->
+                    tts.setOnUtteranceProgressListener(mUtteranceProgressListener)
+                    if (setLocale(tts, LinkedList(listOf(
+                                    Locale(Localization.getCurrentLocale()),
+                                    Locale.getDefault(),
+                                    Locale.ENGLISH)))) {
+                        // tts initialization completed.
+                        mInitialized = true
+                        speak(context, text)
+                    }
+                }
+            } else {
+                mTTSCallback?.initFailed()
+            }
+        })
+    }
+
+    /**
+     * Sets a TTS language from the given list of locale starting from the first Locale.
+     *
+     * Returns a boolean indicating whether we TTS language is ready to use.
+     */
+    private fun setLocale(tts: TextToSpeech, localeList: LinkedList<Locale>): Boolean {
+        if (localeList.isEmpty()) {
+            return false
+        }
+        val locale = localeList.pop()
+        return when (tts.isLanguageAvailable(locale)) {
+            TextToSpeech.LANG_AVAILABLE,
+            TextToSpeech.LANG_COUNTRY_AVAILABLE,
+            TextToSpeech.LANG_COUNTRY_VAR_AVAILABLE -> {
+                // Set language
+                tts.language = locale
+
+                // Check if voice data is present or not.
+                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
+                    if (tts.voice != null) {
```

Comment:
should we returning `false` when `tts.voice==null` ?

Reply:
actually no. `tts.voice` can be null for any reason but we know for sure that the language is available just that we won't be sure whether the voice data is available. So if we return false here, that means we're saying we can't use TTS at all, but we can. And if the internet is available, tts will work just fine. And if not, we'll simply say a toast that TTS won't work cuz `utteranceListener` `onError` will get called.

Reply:
I see, thanks for the explanation. Would be good to add an explicit else here and return `true` with a comment so that  we don't make a mistake of returning false from here in future.